File: app/lib/models/store.dart
import 'product.dart';
import 'portfolio_item.dart';
import 'review.dart';

class Store {
  final int id;
  final int ownerId;
  final String name;
  final String? slug;
  final String? description;
  final String? coverImage;
  final String? profileImage;
  final String? address;
  final String? city;
  final String? state;
  final int? wilayaId;
  final int? communeId;
  final String? category;
  final int? categoryId;
  final List<int> categoryIds;
  final int? subcategoryId;
  final List<int> subcategoryIds;
  final bool providesServices;
  final List<int> serviceCategoryIds;
  final String? phone;
  final List<String> phones;
  final String? email;
  final String? website;
  final double? lat;
  final double? lng;
  final String? mapUrl;
  final double rating;
  final int reviewCount;
  final int followerCount;
  final bool isOpen;
  final bool isFeatured;
  final bool isSponsored;
  final Map<String, String> socialLinks;
  final DateTime? createdAt;
  
  // Related data
  final String? ownerName;
  final String? ownerPic;
  final bool isSaved;
  final bool isFollowing;
  final List<Product> products;
  final List<PortfolioItem> portfolio;
  final List<Review> reviews;
  final Map<String, dynamic> businessHours;

  Store({
    required this.id,
    required this.ownerId,
    required this.name,
    this.slug,
    this.description,
    this.coverImage,
    this.profileImage,
    this.address,
    this.city,
    this.state,
    this.wilayaId,
    this.communeId,
    this.category,
    this.categoryId,
    this.categoryIds = const [],
    this.subcategoryId,
    this.subcategoryIds = const [],
    this.providesServices = false,
    this.serviceCategoryIds = const [],
    this.phone,
    this.phones = const [],
    this.email,
    this.website,
    this.lat,
    this.lng,
    this.mapUrl,
    this.rating = 0,
    this.reviewCount = 0,
    this.followerCount = 0,
    this.isOpen = true,
    this.isFeatured = false,
    this.isSponsored = false,
    this.socialLinks = const {},
    this.createdAt,
    this.ownerName,
    this.ownerPic,
    this.isSaved = false,
    this.isFollowing = false,
    this.products = const [],
    this.portfolio = const [],
    this.reviews = const [],
    this.businessHours = const {},
  });

  factory Store.fromJson(Map<String, dynamic> json) {
    // Parse phones
    List<String> phonesList = [];
    if (json['phones'] != null) {
      if (json['phones'] is List) {
        phonesList = List<String>.from(json['phones']);
      }
    }

    // Parse social links
    Map<String, String> socialMap = {};
    if (json['social_links'] != null) {
      if (json['social_links'] is Map) {
        socialMap = Map<String, String>.from(json['social_links']);
      }
    }

    return Store(
      id: json['id'] as int? ?? 0,
      ownerId: json['owner_id'] as int? ?? 0,
      name: json['name'] as String? ?? '',
      slug: json['slug'] as String?,
      description: json['description'] as String?,
      coverImage: json['cover_image'] as String?,
      profileImage: json['profile_image'] as String?,
      address: json['address'] as String?,
      city: json['city'] as String?,
      state: json['state'] as String?,
      wilayaId: json['wilaya_id'] as int?,
      communeId: json['commune_id'] as int?,
      category: json['category'] as String?,
      categoryId: json['category_id'] as int?,
      categoryIds: json['category_ids'] != null
          ? List<int>.from(json['category_ids'])
          : [],
      subcategoryId: json['subcategory_id'] as int?,
      subcategoryIds: json['subcategory_ids'] != null
          ? List<int>.from(json['subcategory_ids'])
          : [],
      providesServices: json['provides_services'] == 1 || json['provides_services'] == true,
      serviceCategoryIds: json['service_category_ids'] != null
          ? List<int>.from(json['service_category_ids'])
          : [],
      phone: json['phone'] as String?,
      phones: phonesList,
      email: json['email'] as String?,
      website: json['website'] as String?,
      lat: json['lat'] != null ? double.tryParse(json['lat'].toString()) : null,
      lng: json['lng'] != null ? double.tryParse(json['lng'].toString()) : null,
      mapUrl: json['map_url'] as String?,
      rating: double.tryParse(json['rating']?.toString() ?? '0') ?? 0,
      reviewCount: json['review_count'] as int? ?? 0,
      followerCount: json['follower_count'] as int? ?? 0,
      isOpen: json['is_open'] == 1 || json['is_open'] == true,
      isFeatured: json['is_featured'] == 1 || json['is_featured'] == true,
      isSponsored: json['is_sponsored'] == 1 || json['is_sponsored'] == true,
      socialLinks: socialMap,
      createdAt: json['created_at'] != null 
          ? DateTime.parse(json['created_at']) 
          : null,
      ownerName: json['owner_name'] as String?,
      ownerPic: json['owner_pic'] as String?,
      isSaved: json['is_saved'] == 1 || json['is_saved'] == true,
      isFollowing: json['is_following'] == 1 || json['is_following'] == true,
      products: json['products'] != null
          ? (json['products'] as List).map((p) => Product.fromJson(p)).toList()
          : [],
      portfolio: _parsePortfolio(json),
      reviews: json['reviews'] != null
          ? (json['reviews'] as List).map((r) => Review.fromJson(r)).toList()
          : [],
      businessHours: json['business_hours'] != null && json['business_hours'] is Map
          ? Map<String, dynamic>.from(json['business_hours'])
          : {},
    );
  }

  /// Helper to parse portfolio items (checks both 'portfolio' and 'portfolio_items' keys)
  static List<PortfolioItem> _parsePortfolio(Map<String, dynamic> json) {
    // Try 'portfolio_items' first (Laravel snake_case from portfolioItems relationship)
    if (json['portfolio_items'] != null && json['portfolio_items'] is List) {
      return (json['portfolio_items'] as List)
          .map((p) => PortfolioItem.fromJson(p))
          .toList();
    }
    // Fallback to 'portfolio'
    if (json['portfolio'] != null && json['portfolio'] is List) {
      return (json['portfolio'] as List)
          .map((p) => PortfolioItem.fromJson(p))
          .toList();
    }
    return [];
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'owner_id': ownerId,
      'name': name,
      'slug': slug,
      'description': description,
      'cover_image': coverImage,
      'profile_image': profileImage,
      'address': address,
      'city': city,
      'state': state,
      'wilaya_id': wilayaId,
      'commune_id': communeId,
      'category': category,
      'category_id': categoryId,
      'category_ids': categoryIds,
      'subcategory_id': subcategoryId,
      'subcategory_ids': subcategoryIds,
      'provides_services': providesServices,
      'service_category_ids': serviceCategoryIds,
      'phone': phone,
      'phones': phones,
      'email': email,
      'website': website,
      'lat': lat,
      'lng': lng,
      'map_url': mapUrl,
      'rating': rating,
      'review_count': reviewCount,
      'follower_count': followerCount,
      'is_open': isOpen,
      'is_featured': isFeatured,
      'is_sponsored': isSponsored,
      'social_links': socialLinks,
      'business_hours': businessHours,
    };
  }

  String get location {
    if (city != null && state != null) {
      return '$city, $state';
    }
    return city ?? state ?? '';
  }

  String get fullAddress {
    final parts = [address, city, state].where((p) => p != null && p.isNotEmpty);
    return parts.join(', ');
  }

  bool get hasLocation => lat != null && lng != null;

  Store copyWith({
    bool? isSaved,
    bool? isFollowing,
    int? followerCount,
    List<Product>? products,
    List<PortfolioItem>? portfolio,
    List<Review>? reviews,
    List<int>? categoryIds,
    List<int>? subcategoryIds,
    bool? providesServices,
    List<int>? serviceCategoryIds,
    Map<String, dynamic>? businessHours,
  }) {
    return Store(
      id: id,
      ownerId: ownerId,
      name: name,
      slug: slug,
      description: description,
      coverImage: coverImage,
      profileImage: profileImage,
      address: address,
      city: city,
      state: state,
      wilayaId: wilayaId,
      communeId: communeId,
      category: category,
      categoryId: categoryId,
      categoryIds: categoryIds ?? this.categoryIds,
      subcategoryId: subcategoryId,
      subcategoryIds: subcategoryIds ?? this.subcategoryIds,
      providesServices: providesServices ?? this.providesServices,
      serviceCategoryIds: serviceCategoryIds ?? this.serviceCategoryIds,
      phone: phone,
      phones: phones,
      email: email,
      website: website,
      lat: lat,
      lng: lng,
      mapUrl: mapUrl,
      rating: rating,
      reviewCount: reviewCount,
      followerCount: followerCount ?? this.followerCount,
      isOpen: isOpen,
      isFeatured: isFeatured,
      isSponsored: isSponsored,
      socialLinks: socialLinks,
      createdAt: createdAt,
      ownerName: ownerName,
      ownerPic: ownerPic,
      isSaved: isSaved ?? this.isSaved,
      isFollowing: isFollowing ?? this.isFollowing,
      products: products ?? this.products,
      portfolio: portfolio ?? this.portfolio,
      reviews: reviews ?? this.reviews,
      businessHours: businessHours ?? this.businessHours,
    );
  }

}

File: api/app/Models/Store.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use App\Models\StoreSection;

class Store extends Model
{
    use HasFactory;

    protected $fillable = [
        'owner_id',
        'name',
        'slug',
        'description',
        'cover_image',
        'profile_image',
        'address',
        'city',
        'state',
        'wilaya_id',
        'commune_id',
        'category',
        'category_id',
        'category_ids',
        'subcategory_id',
        'subcategory_ids',
        'provides_services',
        'service_category_ids',
        'phone',
        'phones',
        'email',
        'website',
        'map_url',
        'lat',
        'lng',
        'rating',
        'review_count',
        'follower_count',
        'is_open',
        'is_featured',
        'is_sponsored',
        'is_validated',
        'social_links',
        'business_hours',
        'postal_code',
        'activities',
    ];


    /**
     * Resolve the route binding for the model.
     *
     * @param  mixed  $value
     * @param  string|null  $field
     * @return \Illuminate\Database\Eloquent\Model|null
     */
    public function resolveRouteBinding($value, $field = null)
    {
        // If the value is numeric, assume it's an ID
        if (is_numeric($value)) {
            return $this->where('id', $value)->firstOrFail();
        }

        // Otherwise, assume it's a slug
        return $this->where('slug', $value)->firstOrFail();
    }


    protected $casts = [
        'phones' => 'array',
        'social_links' => 'array',
        'subcategory_ids' => 'array',
        'category_ids' => 'array',
        'service_category_ids' => 'array',
        'lat' => 'float',
        'lng' => 'float',
        'rating' => 'float',
        'is_open' => 'boolean',
        'is_featured' => 'boolean',
        'is_sponsored' => 'boolean',
        'is_validated' => 'boolean',
        'provides_services' => 'boolean',
        'category_id' => 'integer',
        'subcategory_id' => 'integer',
        'wilaya_id' => 'integer',
        'commune_id' => 'integer',
        'business_hours' => 'array',
        // 'activities' is stored as simple text (comma separated), no cast needed unless JSON but user said concat.
        // Wait user said "insert the values as concatination". So it is a string.
    ];

    /**
     * Get the category of the store.
     */
    public function categoryRelation()
    {
        return $this->belongsTo(Category::class, 'category_id');
    }

    /**
     * Get the subcategory of the store.
     */
    public function subcategoryRelation()
    {
        return $this->belongsTo(Category::class, 'subcategory_id');
    }

    /**
     * Get the wilaya (state) of the store.
     */
    public function wilaya()
    {
        return $this->belongsTo(Wilaya::class);
    }

    /**
     * Get the commune (city) of the store.
     */
    public function commune()
    {
        return $this->belongsTo(Commune::class);
    }

    /**
     * Get the owner of the store.
     */
    public function owner()
    {
        return $this->belongsTo(User::class, 'owner_id');
    }

    /**
     * Get the products for this store.
     */
    public function products()
    {
        return $this->hasMany(Product::class);
    }

    /**
     * Get the portfolio items for this store.
     */
    public function portfolioItems()
    {
        return $this->hasMany(PortfolioItem::class);
    }

    /**
     * Get the reviews for this store.
     */
    public function reviews()
    {
        return $this->hasMany(Review::class);
    }

    /**
     * Get the customizable sections for this store.
     */
    public function sections()
    {
        return $this->hasMany(StoreSection::class)->orderBy('sort_order');
    }

    /**
     * Get users who saved this store.
     */
    public function savedByUsers()
    {
        return $this->belongsToMany(User::class, 'saved_stores', 'store_id', 'user_id')
            ->withTimestamps();
    }

    /**
     * Get users who follow this store.
     */
    public function followers()
    {
        return $this->belongsToMany(User::class, 'store_followers', 'store_id', 'user_id')
            ->withTimestamps();
    }

    /**
     * Get the location string.
     */
    public function getLocationAttribute(): string
    {
        return collect([$this->city, $this->state])
            ->filter()
            ->implode(', ');
    }

    /**
     * Scope for featured stores.
     */
    public function scopeFeatured($query)
    {
        return $query->where('is_featured', true);
    }

    /**
     * Scope for sponsored stores.
     */
    public function scopeSponsored($query)
    {
        return $query->where('is_sponsored', true);
    }

    /**
     * Check if the store is saved by the currently authenticated user.
     */
    public function getIsSavedAttribute(): bool
    {
        $user = auth('sanctum')->user();

        if (!$user) {
            return false;
        }

        return $this->savedByUsers()->where('user_id', $user->id)->exists();
    }

    /**
     * Check if the store is followed by the currently authenticated user.
     */
    public function getIsFollowingAttribute(): bool
    {
        $user = auth('sanctum')->user();

        if (!$user) {
            return false;
        }

        return $this->followers()->where('user_id', $user->id)->exists();
    }

    /**
     * Append is_saved and is_following to JSON serialization.
     */
    protected $appends = ['is_saved', 'is_following'];

    /**
     * Update rating based on reviews.
     */
    public function updateRating()
    {
        $avgRating = $this->reviews()->avg('rating') ?? 0;
        $reviewCount = $this->reviews()->count();

        $this->update([
            'rating' => round($avgRating, 1),
            'review_count' => $reviewCount,
        ]);
    }

    /**
     * Update follower count.
     */
    public function updateFollowerCount()
    {
        $this->update([
            'follower_count' => $this->followers()->count(),
        ]);
    }
}

